python/pyspark/tests.py

```diff
@@ -1246,6 +1277,25 @@ def test_pipe_unicode(self):
         result = rdd.pipe('cat').collect()
         self.assertEqual(data, result)
 
+    def test_stopiteration_in_client_code(self):
+
+        def stopit(*x):
+            raise StopIteration()
+
+        seq_rdd = self.sc.parallelize(range(10))
+        keyed_rdd = self.sc.parallelize((x % 2, x) for x in range(10))
+        exc = Py4JJavaError, RuntimeError
```

Comment:
Hm .. can we just check one of explicit exception if it's not hard? Py4JJavaError or RuntimeError somehow sounds a bit two arbitrary exceptions ...

Reply:
Both of them can happen, depending on where the `StopIteration` is raised. Consider for example `RDD.reduce`: if the exception is raised when reducing inside a partition, the user will get a `Py4JJavaError`, but if the error is raised when reducing locally the results [here](https://github.com/e-dorigatti/spark/blob/fix_spark_23754/python/pyspark/rdd.py#L858), it will be a `RuntimeError` (the one we raise in `fail_on_stopiteration`)

Reply:
Wait .. just for clarification, one of both exception can be arbitrarily raised for each execution?

Reply:
Well no in the tests it should be deterministic. Do you prefer to catch only the relevant exception for each line?

Reply:
Yea please with a comment for `aggregate`.